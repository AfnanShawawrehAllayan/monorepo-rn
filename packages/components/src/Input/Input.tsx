import React, { useState } from 'react';
import {
  View,
  TextInput,
  StyleSheet,
  TextInputProps,
  ViewStyle,
  TextStyle,
  TouchableOpacity,
} from 'react-native';
import { useScale } from 'hooks';
import { useTheme } from 'theme';
import { Text } from '../Text';

interface InputProps extends TextInputProps {
  label?: string;
  error?: string;
  leftIcon?: React.ReactNode;
  rightIcon?: React.ReactNode;
  containerStyle?: ViewStyle;
  inputStyle?: TextStyle;
  labelStyle?: TextStyle;
  errorStyle?: TextStyle;
  onRightIconPress?: () => void;
  onLeftIconPress?: () => void;
}

export const Input = ({
  label,
  error,
  leftIcon,
  rightIcon,
  containerStyle,
  inputStyle,
  labelStyle,
  errorStyle,
  onRightIconPress,
  onLeftIconPress,
  ...props
}: InputProps) => {
  const theme = useTheme();
  const { scale } = useScale();
  const [isFocused, setIsFocused] = useState(false);

  const getBorderColor = () => {
    if (error) return theme.colors.error;
    if (isFocused) return theme.colors.primary;
    return theme.colors.border;
  };

  const styles = StyleSheet.create({
    container: {
      marginBottom: scale(16),
      width: '100%',
    },
    label: {
      marginBottom: scale(6),
    },
    inputContainer: {
      flexDirection: 'row',
      alignItems: 'center',
      borderWidth: 1,
      borderRadius: scale(8),
      overflow: 'hidden',
    },
    input: {
      flex: 1,
      height: scale(48),
      fontSize: scale(16),
      paddingVertical: scale(8),
    },
    iconContainer: {
      padding: scale(12),
      justifyContent: 'center',
      alignItems: 'center',
    },
    error: {
      marginTop: scale(4),
    },
  });

  return (
    <View style={[styles.container, containerStyle]}>
      {label && (
        <Text 
          variant="body2" 
          style={[
            styles.label, 
            { color: error ? theme.colors.error : theme.colors.text },
            labelStyle
          ]}
        >
          {label}
        </Text>
      )}

      <View 
        style={[
          styles.inputContainer, 
          { 
            borderColor: getBorderColor(),
            backgroundColor: theme.colors.surface,
          }
        ]}
      >
        {leftIcon && (
          <TouchableOpacity 
            onPress={onLeftIconPress} 
            disabled={!onLeftIconPress}
            style={styles.iconContainer}
          >
            {leftIcon}
          </TouchableOpacity>
        )}

        <TextInput
          style={[
            styles.input,
            {
              color: theme.colors.text,
              paddingLeft: leftIcon ? 0 : scale(12),
              paddingRight: rightIcon ? 0 : scale(12),
            },
            inputStyle,
          ]}
          placeholderTextColor={theme.colors.textSecondary}
          onFocus={() => setIsFocused(true)}
          onBlur={() => setIsFocused(false)}
          {...props}
        />

        {rightIcon && (
          <TouchableOpacity 
            onPress={onRightIconPress} 
            disabled={!onRightIconPress}
            style={styles.iconContainer}
          >
            {rightIcon}
          </TouchableOpacity>
        )}
      </View>

      {error && (
        <Text
          variant="caption"
          style={[styles.error, { color: theme.colors.error }, errorStyle]}
        >
          {error}
        </Text>
      )}
    </View>
  );
}; 