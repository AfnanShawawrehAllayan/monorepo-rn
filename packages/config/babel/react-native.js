module.exports = {
  presets: ['@react-native/babel-preset'],
  plugins: [
    [
      'module-resolver',
      {
        root: ['.'],
        extensions: [
          '.ios.ts',
          '.android.ts',
          '.ts',
          '.ios.tsx',
          '.android.tsx',
          '.tsx',
          '.jsx',
          '.js',
          '.json',
        ],
        alias: {
          '@components': './src/components',
          '@screens': './src/screens',
          '@navigation': './src/navigation',
          '@utils': './src/utils',
          '@hooks': './src/hooks',
          '@assets': './src/assets',
          '@config': './src/config',
        },
      },
    ],
    'react-native-reanimated/plugin',
  ],
  env: {
    production: {
      plugins: ['transform-remove-console'],
    },
  },
};
