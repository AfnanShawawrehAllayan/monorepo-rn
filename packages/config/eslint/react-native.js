module.exports = {
  root: true,
  extends: [
    'eslint:recommended',
    'plugin:@typescript-eslint/recommended',
    'plugin:react/recommended',
    'plugin:react-hooks/recommended',
    'plugin:react-native/all',
    'plugin:import/errors',
    'plugin:import/warnings',
    'plugin:import/typescript',
    'plugin:jest/recommended',
    'prettier',
  ],
  parser: '@typescript-eslint/parser',
  parserOptions: {
    ecmaFeatures: {
      jsx: true,
    },
    ecmaVersion: 2021,
    sourceType: 'module',
  },
  plugins: ['@typescript-eslint', 'react', 'react-native', 'import', 'jest'],
  settings: {
    react: {
      version: 'detect',
    },
    'import/resolver': {
      typescript: {},
      node: {
        extensions: ['.js', '.jsx', '.ts', '.tsx'],
      },
    },
  },
  rules: {
    // React Native specific rules
    'react-native/no-unused-styles': 'error',
    'react-native/split-platform-components': 'error',
    'react-native/no-inline-styles': 'warn',
    'react-native/no-color-literals': 'warn',
    'react-native/no-raw-text': ['warn', { skip: ['Text'] }],
    'react-native/no-single-element-style-arrays': 'error',

    // TypeScript rules
    '@typescript-eslint/explicit-module-boundary-types': 'warn',
    '@typescript-eslint/no-explicit-any': 'warn',
    '@typescript-eslint/no-unused-vars': ['error', { argsIgnorePattern: '^_' }],
    '@typescript-eslint/consistent-type-imports': ['error', { prefer: 'type-imports' }],

    // React rules
    'react/prop-types': 'off',
    'react/react-in-jsx-scope': 'off',
    'react/display-name': 'off',
    'react/jsx-curly-brace-presence': ['error', { props: 'never', children: 'never' }],

    // Import rules
    'import/order': [
      'error',
      {
        groups: ['builtin', 'external', 'internal', 'parent', 'sibling', 'index'],
        'newlines-between': 'always',
        alphabetize: {
          order: 'asc',
          caseInsensitive: true,
        },
      },
    ],
    'import/no-duplicates': 'error',
    'import/no-unresolved': 'error',
    'import/no-cycle': 'error',

    // Jest rules
    'jest/no-disabled-tests': 'warn',
    'jest/no-focused-tests': 'error',
    'jest/no-identical-title': 'error',
    'jest/valid-expect': 'error',
  },
  env: {
    'react-native/react-native': true,
    jest: true,
    'jest/globals': true,
  },
  overrides: [
    {
      files: ['*.test.ts', '*.test.tsx', '*.spec.ts', '*.spec.tsx'],
      rules: {
        '@typescript-eslint/no-explicit-any': 'off',
        'react-native/no-raw-text': 'off',
      },
    },
    {
      files: ['*.js', '*.jsx'],
      rules: {
        '@typescript-eslint/explicit-module-boundary-types': 'off',
      },
    },
  ],
};
