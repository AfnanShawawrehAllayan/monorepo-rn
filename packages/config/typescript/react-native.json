{
  "$schema": "https://json.schemastore.org/tsconfig",
  "display": "React Native",
  "compilerOptions": {
    "target": "esnext",
    "module": "esnext",
    "lib": ["esnext"],
    "allowJs": true,
    "jsx": "react-native",
    "noEmit": true,
    "isolatedModules": true,
    "strict": true,
    "moduleResolution": "node",
    "allowSyntheticDefaultImports": true,
    "esModuleInterop": true,
    "skipLibCheck": true,
    "resolveJsonModule": true,
    "baseUrl": ".",
    "paths": {
      "*": ["*", "*.ios", "*.android"]
    },
    "forceConsistentCasingInFileNames": true,
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "noImplicitReturns": true,
    "noFallthroughCasesInSwitch": true,
    "noUncheckedIndexedAccess": true,
    "noImplicitOverride": true,
    "noPropertyAccessFromIndexSignature": true,
    "exactOptionalPropertyTypes": true,
    "types": ["jest", "react-native"]
  },
  "exclude": ["node_modules", "babel.config.js", "metro.config.js", "jest.config.js"]
}
