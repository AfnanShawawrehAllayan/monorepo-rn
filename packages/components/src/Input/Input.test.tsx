// Mock modules before imports
jest.mock('react-native/Libraries/Utilities/Dimensions', () => ({
  get: jest.fn().mockReturnValue({ width: 375, height: 812 }),
}));

jest.mock('react-native', () => {
  const RN = jest.requireActual('react-native');
  RN.PixelRatio = {
    ...RN.PixelRatio,
    roundToNearestPixel: jest.fn((value) => value),
    get: jest.fn(() => 2),
  };
  return RN;
});

jest.mock('hooks', () => ({
  useScale: () => ({
    scale: jest.fn((value) => value),
    verticalScale: jest.fn((value) => value),
    moderateScale: jest.fn((value) => value),
  }),
}));

// Mock theme
jest.mock('theme', () => ({
  useTheme: () => ({
    colors: {
      primary: '#007AFF',
      secondary: '#5856D6',
      background: '#FFFFFF',
      surface: '#F5F5F5',
      text: '#000000',
      textSecondary: '#666666',
      textOnPrimary: '#FFFFFF',
      textDisabled: '#999999',
      disabled: '#999999',
      error: '#FF3B30',
      success: '#34C759',
      border: '#E0E0E0',
    },
    spacing: {
      xs: 4,
      s: 8,
      m: 16,
      l: 24,
      xl: 32,
    },
    typography: {
      fontSize: {
        h1: 32,
        h2: 28,
        h3: 24,
        h4: 20,
        h5: 18,
        h6: 16,
        body1: 16,
        body2: 14,
        caption: 12,
        button: 16,
      },
    },
  }),
}));

import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import { Input } from './Input';
import { View } from 'react-native';

describe('Input Component', () => {
  it('renders correctly with default props', () => {
    const { getByPlaceholderText, toJSON } = render(
      <Input placeholder="Enter text" />
    );
    expect(getByPlaceholderText('Enter text')).toBeTruthy();
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders with label', () => {
    const { getByText, toJSON } = render(
      <Input label="Username" placeholder="Enter username" />
    );
    expect(getByText('Username')).toBeTruthy();
    expect(toJSON()).toMatchSnapshot();
  });

  it('shows error message when provided', () => {
    const errorMessage = 'This field is required';
    const { getByText, toJSON } = render(
      <Input error={errorMessage} placeholder="Enter text" />
    );
    expect(getByText(errorMessage)).toBeTruthy();
    expect(toJSON()).toMatchSnapshot();
  });

  it('handles text input correctly', () => {
    const onChangeText = jest.fn();
    const { getByPlaceholderText } = render(
      <Input placeholder="Enter text" onChangeText={onChangeText} />
    );
    
    const input = getByPlaceholderText('Enter text');
    fireEvent.changeText(input, 'Hello World');
    expect(onChangeText).toHaveBeenCalledWith('Hello World');
  });

  it('handles focus and blur events', () => {
    const onFocus = jest.fn();
    const onBlur = jest.fn();
    const { getByPlaceholderText } = render(
      <Input
        placeholder="Enter text"
        onFocus={onFocus}
        onBlur={onBlur}
      />
    );
    
    const input = getByPlaceholderText('Enter text');
    fireEvent(input, 'focus');
    expect(onFocus).toHaveBeenCalled();
    
    fireEvent(input, 'blur');
    expect(onBlur).toHaveBeenCalled();
  });

  it('renders with left and right icons', () => {
    const leftIconTestId = 'left-icon';
    const rightIconTestId = 'right-icon';
    const { getByTestId, toJSON } = render(
      <Input
        placeholder="Enter text"
        leftIcon={<View testID={leftIconTestId} />}
        rightIcon={<View testID={rightIconTestId} />}
      />
    );
    
    expect(getByTestId(leftIconTestId)).toBeTruthy();
    expect(getByTestId(rightIconTestId)).toBeTruthy();
    expect(toJSON()).toMatchSnapshot();
  });

  it('handles icon press events', () => {
    const onLeftIconPress = jest.fn();
    const onRightIconPress = jest.fn();
    const { getByTestId } = render(
      <Input
        placeholder="Enter text"
        leftIcon={<View testID="left-icon" />}
        rightIcon={<View testID="right-icon" />}
        onLeftIconPress={onLeftIconPress}
        onRightIconPress={onRightIconPress}
      />
    );
    
    const leftIcon = getByTestId('left-icon');
    const rightIcon = getByTestId('right-icon');
    
    if (leftIcon.parent) {
      fireEvent.press(leftIcon.parent);
      expect(onLeftIconPress).toHaveBeenCalled();
    }
    
    if (rightIcon.parent) {
      fireEvent.press(rightIcon.parent);
      expect(onRightIconPress).toHaveBeenCalled();
    }
  });

  it('applies custom styles', () => {
    const containerStyle = { marginTop: 20 };
    const inputStyle = { fontSize: 18 };
    const labelStyle = { fontSize: 16 };
    const errorStyle = { fontSize: 12 };
    
    const { getByTestId, toJSON } = render(
      <Input
        testID="custom-input"
        placeholder="Enter text"
        containerStyle={containerStyle}
        inputStyle={inputStyle}
        labelStyle={labelStyle}
        errorStyle={errorStyle}
        label="Custom Input"
        error="Error message"
      />
    );
    
    expect(toJSON()).toMatchSnapshot();
  });
}); 
