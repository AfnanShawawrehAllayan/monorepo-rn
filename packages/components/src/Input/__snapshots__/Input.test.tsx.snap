// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input Component applies custom styles 1`] = `
<View
  style={
    [
      {
        "marginBottom": 16,
        "width": "100%",
      },
      {
        "marginTop": 20,
      },
    ]
  }
>
  <Text
    style={
      [
        {
          "color": "#000000",
          "fontSize": 14,
          "fontWeight": "normal",
          "lineHeight": 21,
          "textAlign": undefined,
        },
        [
          {
            "marginBottom": 6,
          },
          {
            "color": "#FF3B30",
          },
          {
            "fontSize": 16,
          },
        ],
      ]
    }
  >
    Custom Input
  </Text>
  <View
    style={
      [
        {
          "alignItems": "center",
          "borderRadius": 8,
          "borderWidth": 1,
          "flexDirection": "row",
          "overflow": "hidden",
        },
        {
          "backgroundColor": "#F5F5F5",
          "borderColor": "#FF3B30",
        },
      ]
    }
  >
    <TextInput
      onBlur={[Function]}
      onFocus={[Function]}
      placeholder="Enter text"
      placeholderTextColor="#666666"
      style={
        [
          {
            "flex": 1,
            "fontSize": 16,
            "height": 48,
            "paddingVertical": 8,
          },
          {
            "color": "#000000",
            "paddingLeft": 12,
            "paddingRight": 12,
          },
          {
            "fontSize": 18,
          },
        ]
      }
      testID="custom-input"
    />
  </View>
  <Text
    style={
      [
        {
          "color": "#000000",
          "fontSize": 12,
          "fontWeight": "normal",
          "lineHeight": 18,
          "textAlign": undefined,
        },
        [
          {
            "marginTop": 4,
          },
          {
            "color": "#FF3B30",
          },
          {
            "fontSize": 12,
          },
        ],
      ]
    }
  >
    Error message
  </Text>
</View>
`;

exports[`Input Component renders correctly with default props 1`] = `
<View
  style={
    [
      {
        "marginBottom": 16,
        "width": "100%",
      },
      undefined,
    ]
  }
>
  <View
    style={
      [
        {
          "alignItems": "center",
          "borderRadius": 8,
          "borderWidth": 1,
          "flexDirection": "row",
          "overflow": "hidden",
        },
        {
          "backgroundColor": "#F5F5F5",
          "borderColor": "#E0E0E0",
        },
      ]
    }
  >
    <TextInput
      onBlur={[Function]}
      onFocus={[Function]}
      placeholder="Enter text"
      placeholderTextColor="#666666"
      style={
        [
          {
            "flex": 1,
            "fontSize": 16,
            "height": 48,
            "paddingVertical": 8,
          },
          {
            "color": "#000000",
            "paddingLeft": 12,
            "paddingRight": 12,
          },
          undefined,
        ]
      }
    />
  </View>
</View>
`;

exports[`Input Component renders with label 1`] = `
<View
  style={
    [
      {
        "marginBottom": 16,
        "width": "100%",
      },
      undefined,
    ]
  }
>
  <Text
    style={
      [
        {
          "color": "#000000",
          "fontSize": 14,
          "fontWeight": "normal",
          "lineHeight": 21,
          "textAlign": undefined,
        },
        [
          {
            "marginBottom": 6,
          },
          {
            "color": "#000000",
          },
          undefined,
        ],
      ]
    }
  >
    Username
  </Text>
  <View
    style={
      [
        {
          "alignItems": "center",
          "borderRadius": 8,
          "borderWidth": 1,
          "flexDirection": "row",
          "overflow": "hidden",
        },
        {
          "backgroundColor": "#F5F5F5",
          "borderColor": "#E0E0E0",
        },
      ]
    }
  >
    <TextInput
      onBlur={[Function]}
      onFocus={[Function]}
      placeholder="Enter username"
      placeholderTextColor="#666666"
      style={
        [
          {
            "flex": 1,
            "fontSize": 16,
            "height": 48,
            "paddingVertical": 8,
          },
          {
            "color": "#000000",
            "paddingLeft": 12,
            "paddingRight": 12,
          },
          undefined,
        ]
      }
    />
  </View>
</View>
`;

exports[`Input Component renders with left and right icons 1`] = `
<View
  style={
    [
      {
        "marginBottom": 16,
        "width": "100%",
      },
      undefined,
    ]
  }
>
  <View
    style={
      [
        {
          "alignItems": "center",
          "borderRadius": 8,
          "borderWidth": 1,
          "flexDirection": "row",
          "overflow": "hidden",
        },
        {
          "backgroundColor": "#F5F5F5",
          "borderColor": "#E0E0E0",
        },
      ]
    }
  >
    <TouchableOpacity
      disabled={true}
      style={
        {
          "alignItems": "center",
          "justifyContent": "center",
          "padding": 12,
        }
      }
    >
      <View
        testID="left-icon"
      />
    </TouchableOpacity>
    <TextInput
      onBlur={[Function]}
      onFocus={[Function]}
      placeholder="Enter text"
      placeholderTextColor="#666666"
      style={
        [
          {
            "flex": 1,
            "fontSize": 16,
            "height": 48,
            "paddingVertical": 8,
          },
          {
            "color": "#000000",
            "paddingLeft": 0,
            "paddingRight": 0,
          },
          undefined,
        ]
      }
    />
    <TouchableOpacity
      disabled={true}
      style={
        {
          "alignItems": "center",
          "justifyContent": "center",
          "padding": 12,
        }
      }
    >
      <View
        testID="right-icon"
      />
    </TouchableOpacity>
  </View>
</View>
`;

exports[`Input Component shows error message when provided 1`] = `
<View
  style={
    [
      {
        "marginBottom": 16,
        "width": "100%",
      },
      undefined,
    ]
  }
>
  <View
    style={
      [
        {
          "alignItems": "center",
          "borderRadius": 8,
          "borderWidth": 1,
          "flexDirection": "row",
          "overflow": "hidden",
        },
        {
          "backgroundColor": "#F5F5F5",
          "borderColor": "#FF3B30",
        },
      ]
    }
  >
    <TextInput
      onBlur={[Function]}
      onFocus={[Function]}
      placeholder="Enter text"
      placeholderTextColor="#666666"
      style={
        [
          {
            "flex": 1,
            "fontSize": 16,
            "height": 48,
            "paddingVertical": 8,
          },
          {
            "color": "#000000",
            "paddingLeft": 12,
            "paddingRight": 12,
          },
          undefined,
        ]
      }
    />
  </View>
  <Text
    style={
      [
        {
          "color": "#000000",
          "fontSize": 12,
          "fontWeight": "normal",
          "lineHeight": 18,
          "textAlign": undefined,
        },
        [
          {
            "marginTop": 4,
          },
          {
            "color": "#FF3B30",
          },
          undefined,
        ],
      ]
    }
  >
    This field is required
  </Text>
</View>
`;
